lib/auth/authentication_backend.rb

```diff
@@ -176,32 +205,59 @@ def initiate_auth(email:, password:)
     raise AuthenticationBackendException.new(e.message)
   end
 
-  # Returns an authentication response
+  # Returns an authentication response normally with JWT
   def respond_to_challenge(params)
     challenge_name = params[:challenge_name]
+    username = params[:challenge_parameters]['USER_ID_FOR_SRP'] || params[:email]
+    challenge_responses = { "USERNAME": username }
     case challenge_name
     when 'NEW_PASSWORD_REQUIRED'
-      challenge_responses = {
-        "USERNAME": params[:challenge_parameters]['USER_ID_FOR_SRP'],
-        "NEW_PASSWORD": params[:new_password]
-      }
-    when 'SOFTWARE_TOKEN_MFA'
-      challenge_responses = {
-        "USERNAME": params[:challenge_parameters]['USER_ID_FOR_SRP'],
-        "SOFTWARE_TOKEN_MFA_CODE": params[:totp_code]
-      }
+      challenge_responses.merge!("NEW_PASSWORD": params[:new_password])
+      send_challenge(session: params[:cognito_session_id], challenge_name: challenge_name, challenge_responses: challenge_responses)
+    when 'SMS_MFA', 'SOFTWARE_TOKEN_MFA'
+      challenge_responses.merge!('SOFTWARE_TOKEN_MFA_CODE': params[:totp_code])
+      send_challenge(session: params[:cognito_session_id], challenge_name: challenge_name, challenge_responses: challenge_responses)
+    when 'MFA_SETUP'
+      totp_resp = client.verify_software_token(session: params[:cognito_session_id], user_code: params[:totp_code])
+      if totp_resp.status == 'SUCCESS'
+        challenge_responses.merge!('ANSWER': 'SOFTWARE_TOKEN_MFA')
+        send_challenge(session: totp_resp.session, challenge_name: challenge_name, challenge_responses: challenge_responses)
+      else
+        raise AuthenticationBackendException.new("Unknown status returned by cognito when verifying software token.  Status returned: #{totp_resp.status}")
+      end
     else
       raise AuthenticationBackendException.new("Unknown challenge_name returned by cognito.  Challenge name returned: #{challenge_name}")
     end
+  rescue Aws::CognitoIdentityProvider::Errors::EnableSoftwareTokenMFAException,
+         Aws::CognitoIdentityProvider::Errors::InvalidPasswordException => e
+    case e.code
+    when 'InvalidPasswordException'
+      devise_msg = 'malformed_password'
+    when 'EnableSoftwareTokenMFAException'
+      devise_msg = 'code_missmatch'
+    end
+    response_hash = {
+      type: RETRY,
+      email: params[:email],
+      secret_code: params[:secret_code],
+      session_id: params[:cognito_session_id],
+      challenge_name: "#{params[:challenge_name]}_RETRY",
+      challenge_parameters: params[:challenge_parameters],
+      flash_message: { code: e.code, message: e.message, devise_message: devise_msg }
```

Comment:
```suggestion
      flash_message: { code: e.code, message: e.message, devise_message: e.code }
```
and if you then change the en.yml  `code_mismatch: "The one time code you entered did not match what was expected"` to `EnableSoftwareTokenMFAException: "The one time code you entered did not match what was expected"` you can get rid of the case above